src/sys/aio.rs

```diff
@@ -165,6 +158,50 @@ impl<'a> AioCb<'a> {
         x
     }
 
+    /// Remove the inner boxed slice, if any, and return it.
+    ///
+    /// The returned value will be the argument that was passed to
+    /// `from_boxed_slice` when this `AioCb` was created.
+    ///
+    /// It is an error to call this method while the `AioCb` is still in
```

Comment:
Would it be possible to actually make this a state machine so this error condition can be caught at compile time? So use a different type for in-progress `AioCb`s and not-in-progress ones?

Reply:
Hm.  It's conceivable.  I would have to change the signature of methods that change the in_progress state to consume the object and return a new one, like this:
```rust
pub fn read(self) -> result::Result<AioCbInProgress, (Error, AioCb)> {
let p: *mut libc::aiocb = &mut self.aiocb;
    Errno::result(unsafe {
        libc::aio_read(p)
    }).map(|_| {
        AioCbInProgress::from_aiocb(self)
    }).map_err(|e| {
        (self, e)
    })
}
```
I would also have to put a `Box` around `AioCb::aiocb`, because the `libc::aiocb` must have a stable address once it's been passed to the kernel.

But I can't guarantee that the compiler will always catch the problem, because if `lio_listio` returns `EIO`, then some operations may be in-progress but not all.  Which means that `LioCb::aiocbs` would have to become a `Vec` of an `Enum` type.  So you wouldn't be able to catch misuse of an `LioCb` at compile time.

I think I could make such an arrangement work with the downsteam mio-aio and tokio-file crates, but they would likewise have to use `Enum` types, so they wouldn't gain anything.  tokio-file already has a totally different way of ensuring that you don't abuse an in-progress aiocb (based on runtime checks).

Do you think it's worth it?

Reply:
I don't particularly think either way, the idea crossed my mind so I asked. I really don't understand this API very well and I don't plan to spend the time to do so. But I wanted to ask in case you hadn't thought about it and it would be a good idea. Current API LGTM.